validator/client/src/main/java/tech/pegasys/teku/validator/client/BeaconProposerPreparer.java

```diff
@@ -0,0 +1,86 @@
+/*
+ * Copyright 2021 ConsenSys AG.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
+ * an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations under the License.
+ */
+
+package tech.pegasys.teku.validator.client;
+
+import java.util.Optional;
+import java.util.stream.Collectors;
+import org.apache.tuweni.bytes.Bytes32;
+import tech.pegasys.teku.infrastructure.unsigned.UInt64;
+import tech.pegasys.teku.spec.Spec;
+import tech.pegasys.teku.ssz.type.Bytes20;
+import tech.pegasys.teku.validator.api.BeaconPreparableProposer;
+import tech.pegasys.teku.validator.api.ValidatorApiChannel;
+import tech.pegasys.teku.validator.api.ValidatorTimingChannel;
+import tech.pegasys.teku.validator.client.loader.OwnedValidators;
+
+public class BeaconProposerPreparer implements ValidatorTimingChannel {
+  private final ValidatorApiChannel validatorApiChannel;
+  private final ValidatorIndexProvider validatorIndexProvider;
+  private final OwnedValidators validators;
+  private final Spec spec;
+  private final Bytes20 feeRecipient;
+
+  public BeaconProposerPreparer(
+      ValidatorApiChannel validatorApiChannel,
+      ValidatorIndexProvider validatorIndexProvider,
+      Optional<Bytes20> feeRecipient,
+      OwnedValidators validators,
+      Spec spec) {
+    this.validatorApiChannel = validatorApiChannel;
+    this.validatorIndexProvider = validatorIndexProvider;
+    this.validators = validators;
+    this.feeRecipient = feeRecipient.orElse(null);
+    this.spec = spec;
+  }
+
+  @Override
+  public void onSlot(UInt64 slot) {
+    if (spec.atSlot(slot).getExecutionPayloadUtil().isPresent()
+        && slot.mod(spec.getSlotsPerEpoch(slot)).isZero()) {
+      validatorIndexProvider
+          .getValidatorIndices(validators.getPublicKeys())
+          .thenApply(
+              integers ->
+                  integers.stream()
+                      .map(
+                          index ->
+                              new BeaconPreparableProposer(UInt64.valueOf(index), feeRecipient))
+                      .collect(Collectors.toList()))
+          .thenAccept(validatorApiChannel::prepareBeaconProposer)
+          .reportExceptions();
```

Comment:
`.reportExceptions()` probably isn't the right thing here.  That reports everything via the default exception handler which asks users to report it as a bug.

I think we want to report errors via something in `ValidatorLogger` similar to `ValidatorLogger.dutyFailed`.